### PR TITLE
Gaussian naive bayes

### PR DESCRIPTION
**Pull request by rlouf**
I implemented the Gaussian Naive Bayes algorithm. I added tests, a notebook and updated the docs.

### CODE REVIEW

**Comment by rlouf** [comment]
Closes #4

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Extra space after `import`.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Are you using this import?

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Extra blank line between the imports and the class per PEP8 (https://www.python.org/dev/peps/pep-0008/#imports).

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Since your testcases are very explicitly named (yea!), you don't need docstrings.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Could you make this comment more explicit? I'm not sure what you mean by `Berk`.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Were you able to get the latex to render locally? My cursory research suggests we may need to put some prefixes to get Sphinx to pick up on it (http://www.sphinx-doc.org/en/stable/ext/math.html).

**Review by parsing-science — CHANGES_REQUESTED** [review]
@rlouf: This looks like it's going to be really cool! Thanks for being the first to add a model. I did a quick check through your code and will do a more thorough review soon.

Just as a heads up, I tend to care a lot about formatting and PEP8 so a lot of my comments are in that vein.

Two main things:
1) There were some inconsistencies in your coding style that I think could be smoothed over. I left comments to point out examples.
2) I found some PEP8 errors in the code you submitted. I pointed out a few, but it would be helpful if you could run the code through a linter before the next review.

Thanks again!

**Comment by rlouf** [comment]
Thanks! What linter would you recommend? I've used `pyflakes` and `pylint` but they missed most of the things you pointed out.

File: CHANGELOG.md
Comment (rlouf) [reply CHANGELOG.md]: That's correct.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Yes, that's more readable indeed.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: As you can see, I use these comments as dividers in the code (there is a "Functions" section at the bottom). I was planning on implementing Binary and Multinomial Naive Bayes in the same file, so left this comment in the source. I can remove it if you think that's a problem.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: True, I didn't double-check. Having said that, I prefer the following style where name & type are on the same line and explanation on the second line: https://github.com/networkx/networkx/blob/master/networkx/algorithms/shortest_paths/astar.py
I find it less clutered. I'll follow the current conventions for now, and we can alway discuss about docstring style later.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: I was trying to follow Scikit-learn's naming conventions (see the code for [k-means](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/k_means_.py) for instance). I personally find they make sense. I'm aware it's a subjective notion, but I think it would be wise to adopt the same conventions as sk-learn here.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: I wasn't happy with the way this was coded when I first wrote the code, and this comment was only intended to be seen by me. I took it off.

File: tests/models/test_NaiveBayes.py
Comment (rlouf) [reply tests/models/test_NaiveBayes.py]: Are you refering to the "2 blank lines before classes" rule?

**Comment by rlouf** [comment]
Also, not directly related to this: most python packages adopt the `snake_case` naming convention for files. Wouldn't it make sense to do the same here?

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: You were right, I changed the prefix and Latex equations now render ok. The only issue I have with it is that it's very ugly; I'll try to see if we can change that by tweaking the settings.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Wouldn't it make sense to follow [PEP8's recommendations](https://www.python.org/dev/peps/pep-0008/#module-level-dunder-names) here too?

File: tests/models/test_NaiveBayes.py
Comment (rlouf) [reply tests/models/test_NaiveBayes.py]: Same as above. I remove it.

**Comment by rlouf** [comment]
I implemented most of the requested changes. I am not sure the docstrings in the rest of the package follow PEP8. I left mine as they were; I can follow the rest of the package, and we may address this in an another PR.

**Comment by parsing-science** [comment]
First of all, I appreciate your enthusiasm for this project. I'm glad you have a lot of ideas to contribute. 

However, I would ask that you please follow the contributor guidelines (https://github.com/parsing-science/pymc3_models/blob/master/CONTRIBUTING.rst), particularly the section about following existing conventions. I made thoughtful deliberate decisions about style and the code itself so I'd appreciate if you respected those. 

If you want to refactor your code to match the rest of the repo, I'll be happy to take another look at your PR.

**Comment by rlouf** [comment]
Sorry, I wasn't questioning your choices. Seeing the references to scikit-learn and numpy in the README and CONTRIBUTING, I started writing code with those in ming. I realised after the fact that the conventions adopted here were slightly different. Not a problem for me, I was happy to refactor to match the rest of the repo.

I ran pyflakes and pycodestyle on both the model and tests. All I got were unused variables in the declaration of the model, and lines longer than 79 characters---they are all shorter than 100 characters thought.

**Comment by twiecki** [comment]
This looks great! @parsing-science are there any outstanding issues you see?

**Comment by parsing-science** [comment]
I've been busy with some personal stuff in the last few months. I'll try to take a look in the next few weeks.

**Comment by rlouf** [comment]
Is there anything I can do to help? I started implementing other models, but I would like to have this PR merged before making them PR-ready.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: I don't think you need this comment since it's the same as the classname.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Since we're importing the models in the __init__ file, I don't think we need to keep this.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Did you get the math to render? I don't see it when I look on github.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Typo: `conditioned`

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Typo: `independent`

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Can you explicitly call mu=0 and sd=100 in these distributions? I got bitten when the PyMC3 API switched from tau to sd as default so I prefer to be explicit.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Need space between `'nuts'` and `are`

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: This is potentially confusing because it returns the class instance, not the PyMC3 model instance. Could you clarify this to make it clear it's an instance of GaussianNaiveBayes and not pm.Model?

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Since this line is too long, could you format it like this:
```
prob_per_feature =  [
      np.sum(prob_per_sample[:, :, i], axis=0)/len(self.trace['mu']) for i in range(self.num_pred)
]
```

I find that easier to read.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: it's a bit confusing that predict_proba returns something called `predictions` and predict returns `pred`. I would expect predict to return `predictions`.  Any ideas for a clearer name for the variable here?

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Maybe put `representing the accuracy score` to be explicit about what scoring method you use.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Let's move this to another file called something like pymc_models/common/utils.py. That way it will be available to any other models that might need this.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: I don't think this comment is necessary since the method is called `test_fit`.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: Typo: `inferred`

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: I think the docstring is unnecessary since the model name is explicit.

**Review by parsing-science — CHANGES_REQUESTED** [review]
@rlouf: Finally got a chance to get some comments back to you! 

Things are looking really good. I've made some bug fixes so please merge master back into your branch when you get a chance.

Your tests don't pass anymore, and I think you're running into the same issue that I had recently (https://github.com/pymc-devs/pymc3/issues/2985). You probably need to specify the dtype of your `y` in the PyMC3 model.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: It is not necessary, but I think it improves readability when there are several classes in the same file. I can remove it for now, and put it back If we ever add another class (MultinomialNaiveBayes, etc.) ?

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: The math formatting is now correct for Sphinx. I got to render it, but I had to change some parameters in the Sphinx config. I’ll adress this in another PR.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: posterior_prediction?

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [reply pymc3_models/models/NaiveBayes.py]: sound good!

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [reply pymc3_models/models/NaiveBayes.py]: sure!

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Fixed.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Done.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Thanks! There must be a bunch of typos left, I'll proof-read before re-submitting.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Technically, `predict` returns a point estimation of the full posterior distribution and `predict_proba` returns the Bayesian posterior prediction. I get how the current names make sense to people who come from traditional ML, but this may bite us further down the line if we ever introduce loss functions for point estimations.

**Comment by rlouf** [comment]
I'm still waiting for the tests to finish, please wait before reviewing the changes.

**Comment by rlouf** [comment]
All right, I made all the requested changes and the tests now pass (your fix did the trick). I also added a picture of the graphical model corresponding to Naive Bayes for the sake of being fancy. If this looks ok to you I can do a final rebase to clean this up before merging.

File: pymc3_models/models/NaiveBayes.py
Comment (parsing-science) [inline pymc3_models/models/NaiveBayes.py]: Could you format this as I suggested? My linter is complaining when it's on two lines.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: These long lines are inconsistent about where the closing paren is. 
I would prefer: 
```
np.testing.assert_equal(
            np.sign(self.pi),
            np.sign(self.test_GNB.trace['pi'].mean(axis=0))
)
```

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: What do you mean by this `precarious` comment?

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: This is not the most informative. If you're not sure, could you just put in something more like: ```TODO: Figure out how to test the score function```.

Additionally, If there's no test, I would just put use the skip decorator (https://docs.python.org/3/library/unittest.html#unittest.skip): ```@unittest.skip('Not implemented yet')```

There's no reason to run the scoring if there's no test since it just takes up time.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: It looks like this test doesn't test anything?

**Review by parsing-science — CHANGES_REQUESTED** [review]
This looks great! Just a few minor comments. I was able to build the docs locally, and I really like the way the math looks. Thanks for figuring that out. 

Once you merge in master and make these final changes, we should be good to merge.

File: pymc3_models/models/NaiveBayes.py
Comment (rlouf) [reply pymc3_models/models/NaiveBayes.py]: Yes sorry, I totally missed that comment :man_facepalming:

File: tests/models/test_NaiveBayes.py
Comment (rlouf) [reply tests/models/test_NaiveBayes.py]: :+1:

File: tests/models/test_NaiveBayes.py
Comment (rlouf) [reply tests/models/test_NaiveBayes.py]: You're right!

File: tests/models/test_NaiveBayes.py
Comment (rlouf) [reply tests/models/test_NaiveBayes.py]: True, somehow I didn't write the part that actually tests something ><

**Comment by rlouf** [comment]
I (finally) addressed all the issues you raised, rebased my branch onto master and update the date in the changelog. This should be ready to merge if you approve of the changes.

File: tests/models/test_NaiveBayes.py
Comment (parsing-science) [inline tests/models/test_NaiveBayes.py]: I just set up Travis CI to run the unittests and realized I should have been setting the random seed. Could you do that in this file? Like here, https://github.com/parsing-science/pymc3_models/blob/master/tests/models/test_LinearRegression.py#L23.

**Review by parsing-science — APPROVED** [review]
@rlouf: finally had time to get back to this library! Your PR looks good.

I've made a bunch of changes in master though including adding a linter and Travis CI. I think I can merge master into your fork if I pull it down and merge it or you could merge in master yourself. Do you have a preference?

**Comment by rlouf** [comment]
Cool! I’ll rebase my branch against master, make the changes, clean the history a little bit and it should be ready to merge.

**Comment by parsing-science** [comment]
@rlouf: thanks for doing this! It looks like the linting is failing so I stopped the unittests so you can fix that stuff up first.

**Comment by rlouf** [comment]
@parsing-science Thanks! Makes sense, I was not using flake8 as a linter, good thing you enforced that with CI! Will make the necessary changes ASAP.

**Comment by rlouf** [comment]
Done! Waiting for Travis to do its job :)

**Comment by parsing-science** [comment]
Actually I pulled down your branch and ran some of the tests so I'm going to merge this. Thanks!

**Comment by rlouf** [comment]
Thanks! This was a great learning experience, thank you for bearing with me. I hope to be able to contribute more models soon.

**Pull request merged**
